File: Source/Libraries/GSF.SortedTreeStore/Snap/Filters/TimestampSeekFilter.cs
//******************************************************************************************************
//  TimestampSeekFilter.cs - Gbtc
//
//  Copyright © 2014, Grid Protection Alliance.  All Rights Reserved.
//
//  Licensed to the Grid Protection Alliance (GPA) under one or more contributor license agreements. See
//  the NOTICE file distributed with this work for additional information regarding copyright ownership.
//  The GPA licenses this file to you under the MIT License (MIT), the "License"; you may
//  not use this file except in compliance with the License. You may obtain a copy of the License at:
//
//      http://opensource.org/licenses/MIT
//
//  Unless agreed to in writing, the subject software distributed under the License is distributed on an
//  "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. Refer to the
//  License for the specific language governing permissions and limitations.
//
//  Code Modification History:
//  ----------------------------------------------------------------------------------------------------
//  11/09/2013 - Steven E. Chisholm
//       Generated original version of source code. 
//     
//******************************************************************************************************

using System;
using System.Data;
using System.Runtime.CompilerServices;
using GSF.IO;
using GSF.Snap.Types;

namespace GSF.Snap.Filters
{
    public partial class TimestampSeekFilter
    {

        /// <summary>
        /// Creates a filter over a single date range. (Inclusive list)
        /// </summary>
        /// <param name="firstTime">the first time if the query (inclusive)</param>
        /// <param name="lastTime">the last time of the query (inclusive)</param>
        /// <returns></returns>
        public static SeekFilterBase<TKey> CreateFromRange<TKey>(DateTime firstTime, DateTime lastTime)
            where TKey : TimestampPointIDBase<TKey>, new()
        {
            return new FixedRange<TKey>((ulong)firstTime.Ticks, (ulong)lastTime.Ticks);
        }

        /// <summary>
        /// Creates a filter over a single date range. (Inclusive list)
        /// </summary>
        /// <param name="firstTime">the first time if the query (inclusive)</param>
        /// <param name="lastTime">the last time of the query (inclusive)</param>
        /// <returns></returns>
        public static SeekFilterBase<TKey> CreateFromRange<TKey>(ulong firstTime, ulong lastTime)
            where TKey : TimestampPointIDBase<TKey>, new()
        {
            return new FixedRange<TKey>(firstTime, lastTime);
        }

        /// <summary>
        /// Creates a filter over a set of date ranges (Similar to down sampled queries)
        /// </summary>
        /// <param name="firstTime">the first time if the query (inclusive)</param>
        /// <param name="lastTime">the last time of the query (inclusive if contained in the intervals)</param>
        /// <param name="mainInterval">the smallest interval that is exact</param>
        /// <param name="subInterval">the interval that will be parsed. Possible to be rounded</param>
        /// <param name="tolerance">the width of every window</param>
        /// <returns>A <see cref="KeySeekFilterBase{TKey}"/> that will be able to do this parsing</returns>
        /// <remarks>
        /// Example uses. FirstTime = 1/1/2013. LastTime = 1/2/2013. 
        ///               MainInterval = 0.1 seconds. SubInterval = 0.0333333 seconds.
        ///               Tolerance = 0.001 seconds.
        /// </remarks>
        public static SeekFilterBase<TKey> CreateFromIntervalData<TKey>(ulong firstTime, ulong lastTime, ulong mainInterval, ulong subInterval, ulong tolerance)
            where TKey : TimestampPointIDBase<TKey>, new()
        {
            return new IntervalRanges<TKey>(firstTime, lastTime, mainInterval, subInterval, tolerance);
        }

        /// <summary>
        /// Creates a filter over a set of date ranges (Similar to down sampled queries)
        /// </summary>
        /// <param name="firstTime">the first time if the query (inclusive)</param>
        /// <param name="lastTime">the last time of the query (inclusive if contained in the intervals)</param>
        /// <param name="interval">the exact interval</param>
        /// <param name="tolerance">the width of every window</param>
        /// <returns>A <see cref="KeySeekFilterBase{TKey}"/> that will be able to do this parsing</returns>
        /// <remarks>
        /// Example uses. FirstTime = 1/1/2013. LastTime = 1/2/2013. 
        ///               MainInterval = 0.1 seconds. SubInterval = 0.0333333 seconds.
        ///               Tolerance = 0.001 seconds.
        /// </remarks>
        public static SeekFilterBase<TKey> CreateFromIntervalData<TKey>(ulong firstTime, ulong lastTime, ulong interval, ulong tolerance)
            where TKey : TimestampPointIDBase<TKey>, new()
        {
            return new IntervalRanges<TKey>(firstTime, lastTime, interval, interval, tolerance);
        }

        /// <summary>
        /// Creates a filter over a single date range. (Inclusive list)
        /// </summary>
        /// <param name="firstTime">the first time if the query (inclusive)</param>
        /// <param name="lastTime">the last time of the query (inclusive)</param>
        /// <param name="mainInterval">the smallest interval that is exact</param>
        /// <param name="subInterval">the interval that will be parsed. Possible to be rounded</param>
        /// <param name="tolerance">the width of every window</param>
        /// <returns>A <see cref="KeySeekFilterBase{TKey}"/> that will be able to do this parsing</returns>
        /// <remarks>
        /// Example uses. FirstTime = 1/1/2013. LastTime = 1/2/2013. 
        ///               MainInterval = 0.1 seconds. SubInterval = 0.0333333 seconds.
        ///               Tolerance = 0.001 seconds.
        /// </remarks>
        public static SeekFilterBase<TKey> CreateFromIntervalData<TKey>(DateTime firstTime, DateTime lastTime, TimeSpan mainInterval, TimeSpan subInterval, TimeSpan tolerance)
            where TKey : TimestampPointIDBase<TKey>, new()
        {
            return new IntervalRanges<TKey>((ulong)firstTime.Ticks, (ulong)lastTime.Ticks, (ulong)mainInterval.Ticks, (ulong)subInterval.Ticks, (ulong)tolerance.Ticks);
        }

        /// <summary>
        /// Creates a filter over a single date range. (Inclusive list)
        /// </summary>
        /// <param name="firstTime">the first time if the query (inclusive)</param>
        /// <param name="lastTime">the last time of the query (inclusive)</param>
        /// <param name="interval">the exact interval to do the scan</param>
        /// <param name="tolerance">the width of every window</param>
        /// <returns>A <see cref="KeySeekFilterBase{TKey}"/> that will be able to do this parsing</returns>
        /// <remarks>
        /// Example uses. FirstTime = 1/1/2013. LastTime = 1/2/2013. 
        ///               Interval = 0.1 seconds.
        ///               Tolerance = 0.001 seconds.
        /// </remarks>
        public static SeekFilterBase<TKey> CreateFromIntervalData<TKey>(DateTime firstTime, DateTime lastTime, TimeSpan interval, TimeSpan tolerance)
            where TKey : TimestampPointIDBase<TKey>, new()
        {
            return new IntervalRanges<TKey>((ulong)firstTime.Ticks, (ulong)lastTime.Ticks, (ulong)interval.Ticks, (ulong)interval.Ticks, (ulong)tolerance.Ticks);
        }

        /// <summary>
        /// Loads a <see cref="KeySeekFilterBase{TKey}"/> from the provided <see cref="stream"/>.
        /// </summary>
        /// <param name="stream">The stream to load the filter from</param>
        /// <returns></returns>
        [MethodImpl(MethodImplOptions.NoOptimization)]
        private static SeekFilterBase<TKey> CreateFromStream<TKey>(BinaryStreamBase stream)
            where TKey : TimestampPointIDBase<TKey>, new()
        {
            byte version = stream.ReadUInt8();
            switch (version)
            {
                case 0:
                    return null;
                case 1:
                    return new FixedRange<TKey>(stream);
                case 2:
                    return new IntervalRanges<TKey>(stream);
                default:
                    throw new VersionNotFoundException("Unknown Version");
            }
        }


    }
}

File: Source/Libraries/PythonAPIModule/snapDB/timestampSeekFilter.py
#******************************************************************************************************
#  timestampSeekFilter.py - Gbtc
#
#  Copyright © 2021, Grid Protection Alliance.  All Rights Reserved.
#
#  Licensed to the Grid Protection Alliance (GPA) under one or more contributor license agreements. See
#  the NOTICE file distributed with this work for additional information regarding copyright ownership.
#  The GPA licenses this file to you under the MIT License (MIT), the "License"; you may not use this
#  file except in compliance with the License. You may obtain a copy of the License at:
#
#      http://opensource.org/licenses/MIT
#
#  Unless agreed to in writing, the subject software distributed under the License is distributed on an
#  "AS-IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. Refer to the
#  License for the specific language governing permissions and limitations.
#
#  Code Modification History:
#  ----------------------------------------------------------------------------------------------------
#  02/06/2021 - J. Ritchie Carroll
#       Generated original version of source code.
#
#******************************************************************************************************

from .seekFilterBase import seekFilterBase
from gsf.binaryStream import binaryStream
from gsf import Ticks
from typing import Optional
from datetime import datetime, timedelta
from uuid import UUID
import numpy as np

class timestampSeekFilter(seekFilterBase):
    """
    Creates a seek filter based on time ranges and optional intervals.
    """

    SnapTypeID = UUID("0f0f9478-dc42-4eef-9f26-231a942ef1fa")
    AllTimes: Optional[seekFilterBase] = None

    def __init__(self, firstTime: np.uint64, lastTime: np.uint64, mainInterval: np.uint64 = 0, subInterval: np.uint64 = 0, tolerance: np.uint64 = 0):
        """
        Creates a seek filter over a single date range (inclusive list).
        
        Parameters
        ----------
        firstTime: the first time if the query (inclusive).
        lastTime: the last time of the query (inclusive).
        mainInterval: the smallest interval that is exact.
        subInterval: the interval that will be parsed (round possible).
        tolerance: the width of every window.
        """
        self.firstTime = firstTime
        self.lastTime = lastTime
        self.mainInterval = mainInterval
        self.subInterval = subInterval
        self.tolerance = tolerance

    @property
    def TypeID(self) -> UUID:
        """
        Gets the Guid uniquely defining this SNAPdb seek filter type.
        """
        return timestampSeekFilter.SnapTypeID

    def Save(self, stream: binaryStream):
        """
        Serializes the filter to a stream.
        """
        targetFixedRange = self.mainInterval == 0 and self.subInterval == 0 and self.tolerance == 0

        stream.WriteByte(1 if targetFixedRange else 2)
        stream.WriteUInt64(self.firstTime)
        stream.WriteUInt64(self.lastTime)

        if targetFixedRange:
            return

        stream.WriteUInt64(self.mainInterval)
        stream.WriteUInt64(self.subInterval)
        stream.WriteUInt64(self.tolerance)

    @staticmethod
    def CreateFromRange(firstTime: datetime, lastTime: datetime):
        """
        Creates a seek filter over a single date range (inclusive list).
        
        Parameters
        ----------
        firstTime: the first time if the query (inclusive).
        lastTime: the last time of the query (inclusive).
        """
        return timestampSeekFilter(
            Ticks.FromDateTime(firstTime),
            Ticks.FromDateTime(lastTime))

    @staticmethod
    def CreateFromRangeInterval(firstTime: datetime, lastTime: datetime, interval: timedelta,  tolerance: timedelta):
        """
        Creates a seek filter over a single date range (inclusive list), skipping values
        based on specified `timedelta` interval and tolerance.
        
        Parameters
        ----------
        firstTime: the first time if the query (inclusive).
        lastTime: the last time of the query (inclusive).
        interval: the exact interval for the scan. Example: 0.1 seconds.
        tolerance: the width of every window. Example: 0.001 seconds.
        """
        return timestampSeekFilter(
            Ticks.FromDateTime(firstTime),
            Ticks.FromDateTime(lastTime),
            Ticks.FromTimeDelta(interval),
            Ticks.FromTimeDelta(interval),
            Ticks.FromTimeDelta(tolerance))

    @staticmethod
    def CreateFromRangeSubInterval(firstTime: datetime, lastTime: datetime, mainInterval: timedelta, subInterval: timedelta, tolerance: timedelta):
        """
        Creates a seek filter over a single date range (inclusive list), skipping values
        based on specified `timedelta` main /sub intervals and tolerance.
        
        Parameters
        ----------
        firstTime: the first time if the query (inclusive).
        lastTime: the last time of the query (inclusive).
        mainInterval: the smallest interval that is exact. Example: 0.1 seconds.
        subInterval: the interval that will be parsed (round possible). Example: 0.0333333 seconds.
        tolerance: the width of every window. Example: 0.001 seconds.
        """
        return timestampSeekFilter(
            Ticks.FromDateTime(firstTime),
            Ticks.FromDateTime(lastTime),
            Ticks.FromTimeDelta(mainInterval),
            Ticks.FromTimeDelta(subInterval),
            Ticks.FromTimeDelta(tolerance))
